extensions/2.0/Vendor/EXT_mesh_features/schema/featureIdAttribute.schema.json

```diff
@@ -23,20 +23,26 @@
     "extensions": {},
     "extras": {}
   },
-  "not": {
-    "anyOf": [
-      {
+  "oneOf": [
+    {
+      "description": "Explicit feature ID. 'attribute' is required; 'offset' and 'repeat' are disallowed.",
+      "required": [
+        "attribute"
+      ],
+      "not": {
         "required": [
-          "attribute",
-          "offset"
+          "offset",
+          "repeat"
         ]
-      },
-      {
+      }
+    },
+    {
+      "description": "Implicit feature ID. Both 'offset' and 'repeat' are optional; 'attribute' is disallowed.",
+      "not": {
         "required": [
-          "attribute",
-          "repeat"
+          "attribute"
         ]
       }
-    ]
-  }
+    }
+  ]
```

Comment:
actually I'm not sure now that you mention this, if `required = [a, b]` means a AND b must be present, then `not.required` here would give `(NOT a) OR (NOT b)`. but this would be true if you had just `a` correct?

https://json-schema.org/understanding-json-schema/reference/combining.html#not

Reply:
I tried this with a JSON schema validator I have as a Vim plugin, indeed, this only rules out the `a AND b` case:

![image](https://user-images.githubusercontent.com/8422414/137021857-604bcb2d-78e9-47a2-85e2-927f2422c706.png)

![image](https://user-images.githubusercontent.com/8422414/137021890-370908d2-23a8-4aab-9dad-cfaef352ffb4.png)


Schema: 
![image](https://user-images.githubusercontent.com/8422414/137022014-6888525a-de07-4c62-9b27-a4531a5c3563.png)

Reply:
Hm, does the previous syntax (with `not.anyOf.required`) do the correct thing here?

I guess we could (a) stick with the previous syntax, or (b) expand this to:

```jsonc
{
      "description": "Explicit feature ID. 'attribute' is required; 'offset' and 'repeat' are disallowed.",
      "required": [
        "attribute"
      ],
      "allOf": [
        {"not": {"required": ["offset"]}},
        {"not": {"required": ["repeat"]}}
      ]
},
```

... would be curious what is clearer from your perspective? I just found `not.anyOf.required` difficult to read; in both cases it is a little unintuitive to think of "required" as meaning present or allowed.

Reply:
It's a bit difficult to stay up to date with the JSON Schema draft versions, and admittedly, I didn't dive into the details of type composition (with nested `not`/`allOf`/`anyOf`) until now. 

Stumbling my way of trying to understand this:

The original structure was 
```
not {
  anyOf[
     required: [attribute, offset],
     required: [attribute, repeat]
  ]
}
```
Which "intuitively" would mean:  When an object has (`attribute` and `offset`) **or** (`attribute` and `repeat`), then it is **not** accepted. The goal is to rule out objects that have `attribute` and one of the other properties. 

The new structure is
```
oneOf[
  {       required: [attribute], not { required: [offset, repeat] } }
  { not { required: [attribute] } }
}
```
Which "intuitively" means: When an object has `attribute`, and **not** (`offset` and `repeat`), then it **is** accepted (when it does **not** have `attribute`, it is also accepted). 

The first part is probably what Peter referred to: If the object has `attribute` and *either* `offset` or `repeat` (but not both), then it will still be accepted. 

(And to my understanding, the suggestion in your last comment should indeed solve this)

---

Regardless of which of the paths is taken (restoring the original one, or using the last one): The **`description`** adds the real value here: A single sentence that can safe others from browsing through the JSON Schema docs and scriblling down truth tables and DeMorgan rules. 

---

Somewhat unrelated: I looked at http://json-schema.org/understanding-json-schema/reference/object.html#required and related docs. There does not seem to be an *explicit* statement that the strings in the `required` array are *implicitly* assumed to be properties. I wondered whether something like 
```
{ 
    required: [ "a", "b" ]
}
```
would have to be written elaborately as
{ 
  "properties": {
    "a": {  ... },
    "b": {  ... }
  },    
  required: [ "a", "b" ]
}
```
but of course I hope that this is **not** the case... that would be awfully repetitive...

Reply:
Based on the previous discussion (and based on https://www.jsonschemavalidator.net/ , FWIW), this appears to be correct.

Actually, I wanted to add a fix for this part (as discussed in slack), but hesitated and went back and forth between the original version and the one that you now used. The verbosity was the main reason for the hesitation. But I think that it is much easier to understand (and *document*) when both cases are modeled *independently*: In the old version, they had been rather "implicit" and it wasn't sensibly possible to document the `anyOf`-elements with a `description`, because they had been contained in the `not`... 

So +1 for that update.